packages/spec/spec_reporter.savi

```diff
@@ -48,13 +48,20 @@
     if (skip) (@env.err.print(" SKIP") | @env.err.print(" OK"))
 
   :fun ref event(spec String, example String, event SpecEventAssert)
+    // TODO: We shouldn't need this variable, but using `event.assert` in the `if` fails compilation.
+    assert = event.assert
+
     if event.success (
       @env.err.write(".")
     |
-      @env.err.write("\n")
-      @env.err.write(Inspect[event.pos.row + 1])
-      @env.err.write(": FAIL: ")
-      @env.err.print(event.pos.string)
+      if (assert <: Assert2) (
+        assert.report_failure(@env)
+      |
+        @env.err.write("\n")
+        @env.err.write(Inspect[event.pos.row + 1])
+        @env.err.write(": FAIL: ")
+        @env.err.print(event.pos.string)
+      )
```

Comment:
As I wrote in a response to your related comment in #102, this restriction is not strictly a bug in the compiler but rather a limitation that we need to deal with fairly regularly.

The way you've dealt with it here does work, but I could also suggest another approach to consider for avoiding the use of an intermediate variable. This approach instead uses `as!` to coerce the type within a `try` block, then falls back to the other block if the error is raised:

```savi
    if event.success (
      @env.err.write(".")
    |
      try (
        event.assert.as!(Assert2).report_failure(@env)
      |
        @env.err.write("\n")
        @env.err.write(Inspect[event.pos.row + 1])
        @env.err.write(": FAIL: ")
        @env.err.print(event.pos.string)
      )
```

Unlike in Ruby or many other languages, this is considered good practice in Savi because we can clearly see and prove to ourselves that the only kind of error that can possibly happen in the `try` block is this conversion error. So we can treat this as a reasonable away to try to do something with the type-casted value, or fall back to another approach if the type-cast didn't work.

In Savi all error-raising call sites are explicitly identifiable with the `!` suffix on them (and red syntax highlighting in the official set of syntax highlighters), so you can cleanly use `try` for control flow with a fallback plan, without worrying about other sources of errors causing you to take an unexpected path to the fallback.